fq.go

```diff
@@ -0,0 +1,172 @@
+package fq
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// FQScheduler is a fair queuing implementation designed for the kube-apiserver.
+// FQ is designed for
+// 1) dispatching requests to be served rather than packets to be transmitted
+// 2) serving multiple requests at once
+// 3) accounting for unknown and varying service time
+type FQScheduler struct {
+	lock         sync.Mutex
+	queues       []*Queue
+	clock        clock.Clock
+	vt           float64
+	C            float64
+	G            float64
+	lastRealTime time.Time
+	robinidx     int
+}
+
+func (q *FQScheduler) chooseQueue(packet *Packet) *Queue {
+	if packet.queueidx < 0 || packet.queueidx > len(q.queues) {
+		panic("no matching queue for packet")
+	}
+	return q.queues[packet.queueidx]
+}
+
+func NewFQScheduler(queues []*Queue, clock clock.Clock) *FQScheduler {
+	fq := &FQScheduler{
+		queues: queues,
+		clock:  clock,
+		vt:     0,
```

Comment:
I think `lastRealTime` needs to be initialized too.

Reply:
Updated, initialized it to the time of initialization: q.clock.Now()